# Basic Information:
### Title:  px_uploader.py does not use monotonic clocks #20311 
### Issue status: Closed
### Author: fkaiser
### Issue open time: Sep 27, 2022
### Fixed by: #20352
# Report
### Report author: fkaiser
### Report Time: Sep 27, 2022
### Report Content:   
Describe the bug  
If I use the Pythons script px_uploader.py to flash my board, the erasing part of the flashing sometimes times-out:  
I noticed that the time-outs always occur when my system clock gets updated after I have called the flashing script e.g. I start the flashing on system time 2022-09-20 08:00:00 and then the system clock gets updated to 2022-09-26 08:10:00 e.g. due to a network connection. The reason seems to be that the script px_uploader.py uses the time() functionality to compute the time difference for the time-out:  
To prevent this failure couldn't we use the monotonic clock of the Python's time module if is available as also recommended by PEP 418:  
However, the monotonic clock of the time module is only available as of Python 3.3 and only always system-wide available as of Python 3.5:  
To Reproduce  
Steps to reproduce the behavior:    
To reproduce the issue, one needs to adapt the system clock during flashing. If you are running a Linux system with the the service timedatectl you can do the following to provoke the error during flashing. (Please be careful with this step!)  
- Build the code  
    
```bash     
 make <target>.px4    
```  
- Check current time  
    
```bash     
 timedatectl     
```  
- Disable auto time sync:  (Revert this change at the end of the test)  
    
```bash     
 sudo timedatectl set-ntp 0    
```  
- Start flashing  
    
```bash     
 python3 px_uploader.py --port <target port> --baud-bootloader <target baudrate> <target>.px4    
```  
- Change time during deletion step of flashing:  
    
```bash     
 sudo timedatectl set-time 2022-12-05    
```  
- See error  
Expected behavior  
The time-outs should be independend of the update of the system time.  
Log Files and Screenshots  

# Comment
## Comment1
### Comment author: dagar
### Comment Time: Sep 27, 2022
### Comment Content:   
A would say a good baseline for minimum python version is probably ~3.6 at this point (Ubuntu 18.04). Python 3.5 was released in 2015.  
Could you open a pull request updating the uploader script?  

## Comment2
### Comment author: fkaiser
### Comment Time: Oct 4, 2022
### Comment Content:   
    
Could you open a pull request updating the uploader script?    
PR that fixes this issue would be here \\\#20352  

## Comment3
### Comment author: junwoo091400
### Comment Time: Oct 6, 2022
### Comment Content:   
Ah good catch, thanks for the report! I will comment in the PR  
